doc/api/esm.md

```diff
@@ -313,50 +275,349 @@ If a package has no exports, setting `"exports": false` can be used instead of
+versions of the package may get used within an application; for example, the
+user’s application code could `import` the ES module version while a dependency
+`require`s the CommonJS version. If that were to occur, two copies of the
+package would be loaded in memory and therefore two separate states would be
+present. This would likely cause hard-to-troubleshoot bugs.
+
+Aside from writing a stateless package (if JavaScript’s `Math` were a package,
+for example, it would be stateless as all of its methods are static), there are
+three ways to isolate state so that it’s shared between the potentially loaded
+CommonJS and ES module instances of the package:
+
+1. If possible, contain all state within an instantiated object. JavaScript’s
+   `Date`, for example, needs to be instantiated to contain state; if it were a
+   package, it would be used like this:
+
+    ```js
+    import date from 'date';
+    const someDate = new date();
+    // someDate contains state; date does not
+    ```
+
+   The `new` keyword isn’t required; a package’s function can return a new
+   object, or modify a passed-in object, to keep the state external to the
+   package.
+
+1. Isolate the state in one or more CommonJS files that are shared between the
+   CommonJS and ES module versions of the package. For example, if the CommonJS
+   and ES module entry points are `index.cjs` and `index.mjs`, respectively:
+
+    ```js
+    // ./node_modules/pkg/index.cjs
+    const state = require('./state.cjs');
+    module.exports.state = state;
+    ```
+
+    ```js
+    // ./node_modules/pkg/index.mjs
+    export * as state from './state.cjs';
+    ```
+
+   Even if `pkg` is used via both `require` and `import` in an application (for
+   example, via `import` in application code and via `require` by a dependency)
+   each reference of `pkg` will contain the same state; and modifying that
+   state from either module system will apply to both.
+
+   A package utilizing this pattern would not be usable unmodified in browsers
+   or in other environments that lack support for CommonJS.
+
+1. Write a package where state is stored globally. This is similar to the
+   previous approach, but instead of isolating state within a shared CommonJS
+   file it is attached to the global object, e.g.
+   `globalThis[Symbol.for('pkg@1.2.3')]`.
+
+   This has the disadvantage of polluting the global namespace, but it is
+   compatible with non-CommonJS environments such as browsers.
+
+For packages using this approach, regardless of which state management pattern
+is chosen, an `instanceof` comparison would return `false` when comparing the
+CommonJS and ES module versions of the package or of objects instantiated from
+each version. End users of such packages need to be aware of this and avoid
+comparing identity in mixed-module system environments, or check against both
+versions:
+
+```js
+import { createRequire } from 'module';
```

Comment:
I considered this. I agree it’s a better solution, but doesn’t this require the original package author to embed this secret symbol, so that it could be found by dependents hoping to compare? Whereas doing two `instanceof`s doesn’t require any cooperation on the part of the original package.

Or could the dependent doing the comparison, whether another package or the end user, attach the symbol to the two singletons? Though doing so would bring us back to needing `createRequire`, in order to access the CommonJS version.

Reply:
If the original author is choosing to publish dual mode then they should embed the symbol and provide an `instanceOfX` method and document that users should call that method instead of using `instanceof` syntax.

Reply:
> If the original author is choosing to publish dual mode then they should embed the symbol

Sure, we should suggest that, but ideally we’d have a solution that doesn’t require effort on both sides (author and consumer).

Reply:
I was going to say it's too bad JavaScript doesn't allow overloading `instanceof`, then I learned that it can!  This is accomplished with [Symbol.hasInstance](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/hasInstance).  I've created a simple but well tested module which makes it very easily deal with `instanceof` in node.js 6.5.0+ and modern browsers.  See https://github.com/cfware/instanceof-unifier#readme for example of fixing `instanceof` for a dual mode transpiled module, so all effort is removed from the consumer.

Reply:
That’s very clever! I’ll have to give some thought as to how to reduce that into a minimal example for inclusion in the docs. I feel like the docs probably can’t link to random personal GitHub example repos.

So that’s a solution where action is required only on the package author’s side; there a different solution where action is required only on the consumer’s side? Is that consumer’s side solution the example I already have in the docs?

I agree the burden should be on the author, so maybe that’s the only example we include, unless there’s value in both?

Reply:
> That’s very clever! I’ll have to give some thought as to how to reduce that into a minimal example for inclusion in the docs. I feel like the docs probably can’t link to random personal GitHub example repos.

Understandable, I'm not aware of any examples of node.js documentation linking to any user-space modules.  For what it's worth the intent is not to be an example repo, it's published to npm.

> So that’s a solution where action is required only on the package author’s side; there a different solution where action is required only on the consumer’s side? Is that consumer’s side solution the example I already have in the docs?
> 
> I agree the burden should be on the author, so maybe that’s the only example we include, unless there’s value in both?

This document will set expectations for how authors should publish dual mode modules, so my vote is against suggesting alternatives where the author could shift the burden to consumers.